tokio/src/sync/notify.rs

```diff
@@ -612,14 +657,12 @@ impl Future for Notified<'_> {
                     }
 
                     // Safety: called while locked.
-                    unsafe {
-                        (*waiter.get()).waker = Some(cx.waker().clone());
-                    }
+                    let w = unsafe { &mut *waiter.get() };
+
+                    w.waker = Some(cx.waker().clone());
 
                     // Insert the waiter into the linked list
-                    //
-                    // safety: pointers from `UnsafeCell` are never null.
-                    waiters.push_front(unsafe { NonNull::new_unchecked(waiter.get()) });
+                    lists.waiters.push_front(w.into());
```

Comment:
If you store a raw pointer for a long time, the safety rules are more complicated when you have a conversion to reference and back. I would prefer to keep this as-is.